### ISSUE TITLE
Is emacs support planned?

### ISSUE THREAD

**Issue by qrilka** [haskell-tools/haskell-tools#370]

**Comment by nboldi** [comment]
Yes, we also want to have emacs or vim support, but currently we don't have the manpower and the know-how to implement it.

**Comment by fycth** [comment]
Is there any API description/documentation available? I mean, how to communicate to the daemon, and other related details.

**Comment by nboldi** [comment]
Yes, the [protocol description](https://github.com/haskell-tools/haskell-tools/blob/master/documentation/development/haskell-tools-refactoring-protocol.md) and [some thinking on the editor integration](https://github.com/haskell-tools/haskell-tools/blob/master/documentation/development/editor-integration.md) are in the documentation. Also the [Haskell API](https://haskell-tools.github.io/master/api/haskell-tools-daemon-0.4.0.0/Language-Haskell-Tools-Refactor-Daemon.html) of the daemon module and the [reference implementation](https://github.com/nboldi/haskell-tools-atom) can also be a useful resource.

**Comment by hyiltiz** [comment]
This would be so cool to be integrated into Vim and Emacs. I imagine integrating into emacs would be much easy if emacs has some framework that acts as some kind of connector. I main vim so I am not quite familiar with emacs except elisp doesn't get in the way.

**Comment by ljli** [comment]
First of all, great tool, the refactorings work well as far as I tried them.
I wrote a very bare bones integration to evaluate it for personal use. https://gist.github.com/ljli/a481e6d62fc8e67ce93e870eb9db50a7
To use it one should be able to just execute the code in a buffer and then use the ht-refact-* functions on a haskell project provided `ghc-pkg list` returns the correct package db.

What would have to be done to get it into a release ready state:
- Provide a way to atomically undo refactorings
- Intelligent handling for project types (cabal, stack, nix, etc.)
  Currently it tries to use explicit package dbs as returned by ghc-pkg list
- Follow all emacs idioms (defvar, etc.)
- Don't rely on projectile (Easy one)
- Define (good) keybindings
- Probably more

I put it on ice for now due to a few reasons:
- Explicit package db handling doesn't work for more than one project in parallel (I see there were commits recently that could have changed this, I did not test them)
- It hogged a lot of resources in some cases (specifically: cabal2nix: 100% on one core (not a problem) and > 4 GB memory (then the OOM killer got involved))
- It operates directly on files. That might be easier for the backend but introduces versioning concerns in every integration (basically all files have to be saved before doing a refactoring to avoid version splits)
- 2 space indentations get converted to 4 space indentation by some refactorings (extract, if remember correctly) (not a big deal, TBH)

On a more general note:
- It'd be great if there were a comparison to [HaRe](https://github.com/RefactoringTools/HaRe) (Is there one and I didn't find it?)
- I'm not very familiar with [LSP](https://github.com/Microsoft/language-server-protocol) but on cursory inspection it seems to be general enough to support the kind of interaction `haskell-tools` needs. It appears to have some traction with editor integrations underway. Maybe it would pay to adopt it. If there is a reason not to, it would be great if we could document that, too.

**Comment by nboldi** [comment]
First of all, thanks for trying and experimenting with emacs support!

 - Undo refactorings is currently an editor integration feature. The daemon sends the needed data as diffs to the client. If the client implements that feature, it can store these diffs and perform undo operations. If needed, this functionality can be merged into the server.

 - Currently cabal (simple and sandbox) and stack projects are handled. Nix is now not supported by I don't know about any obstacles to that. Some functionality is still missing (for example, reload a whole package when .cabal file changes).

 - For the next release, handling of packages will be improved, I solved a few bugs already.

 - We have an issue with a few of the packages, one of it is cabal2nix. It seems that the template haskell code in it just runs infinitely and grabs all memory. Unfortunately I couldn't yet find the boundary conditions that cause this behavior. The following issue documents this problem: https://github.com/haskell-tools/haskell-tools/issues/299.

 - In my experience most refactoring tools only work well with files that are up-to-date on the disk. I found no easy solution to load Haskell modules from anywhere but the disk. However the daemon could do the refactoring without saving the new source files to the disk, this complicates reloading a bit. For now, I think the best option would be to require all files to be saved before refactoring.

 - Indentation change is done for safety (indentation of the introduced `where` clause could conflict with some of the elements of the definition).

 - I'm planning to do a comparison with HaRe soon.

 - LSP seems promising, I will check it out.

**Comment by ljli** [comment]
Sounds great. Just for clarification: I didn't want to imply criticism of the backend with the items on the first list, they were just meant to summarize the state of the emacs code. The diffs sent by haskell-tools seem sufficient to implement atomic undoing. But having the server take care of it would also be fine.

Regarding the on-disk loading. Do you mean the GHC API has problems handling modules from memory? That would be unfortunate. Leaving that aside, I think currently the server has to be informed of every file change. The notification could just be equipped with a code delta and the server could maintain an in-memory version of the file. That (as a data point) is also how the LSP protocol does things. When the client opens a file it controls the "truth" about the file and is responsible to inform the server of every change and the server won't read or write the file on disk.

**Comment by nboldi** [comment]
Yes, I've done some experimenting with it, and I think I found that the [depanal function](https://downloads.haskell.org/~ghc/8.0.2/docs/html/libraries/ghc-8.0.2/GHC.html#v:depanal) always reloads from the disk. So it should be reimplemented using in-memory buffers with timestamps. But this was some time ago, so I may be wrong or the situation may have been changed since. This would be a nice feature to have, but I don't know how much effort would it take to implement this.

**Comment by alanz** [comment]
FYI, there is currently

- https://github.com/vibhavp/emacs-lsp
- https://github.com/alanz/haskell-lsp

And an example implementation at https://github.com/alanz/haskell-ide-engine/tree/lsp

Also https://github.com/alanz/vscode-hie-server
